File: MinecraftPvpPlugin/src/main/java/me/minecraft/minecraftpvpplugin/Gadget.java
package me.minecraft.minecraftpvpplugin;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static org.bukkit.Bukkit.getServer;

abstract public class Gadget extends ItemStack implements Listener {
    public Material material;
    public String name;
    public long duration; // this will only do any effect when SwitchLike = false

    // Switch Like:
    //     false: player click -> activate() -> delete item -> wait for `duration` s  -> deactivate()
    //     true: player click -> activate() -> player click -> deactivate() -> delete item

    public boolean SwitchLike;

    public List<Player> PlayersUsingGadget = new ArrayList<>();

    protected abstract void onActivate(PlayerInteractEvent event);
    protected abstract void onDeactivate(PlayerInteractEvent event);


    public Gadget(Material material, String name, boolean SwitchLike) {
        this.material = material;
        this.name = name;
        this.SwitchLike = SwitchLike;

        this.setType(material);

        ItemMeta meta = this.getItemMeta();
        meta.setDisplayName(name);
        this.setItemMeta(meta);

        getServer().getPluginManager().registerEvents(this, MinecraftPvpPlugin.instance);
    }

    public Gadget(Material material, String name, long duration) {
        this.material = material;
        this.name = name;
        this.duration = duration;
        this.SwitchLike = false;

        this.setType(material);

        ItemMeta meta = this.getItemMeta();
        meta.setDisplayName(name);
        this.setItemMeta(meta);

        System.out.println(getServer().getPluginManager());
        System.out.println(getServer().getPluginManager().getPlugin("MinecraftPvpPlugin"));

        getServer().getPluginManager().registerEvents(this, MinecraftPvpPlugin.instance);
    }

    @EventHandler
    public void Click(PlayerInteractEvent event){
        Player player = event.getPlayer();
        ItemStack item = event.getItem();

        if (event.getItem() != null
                && MinecraftPvpPlugin.IsInPvp(player)
                && Objects.equals(item.getItemMeta().getDisplayName(), name)
                && event.getAction() != Action.PHYSICAL
        ) {
            if (SwitchLike) {
                if (PlayersUsingGadget.contains(player)) {
                    RemoveOneItemInHand(player);
                    PlayersUsingGadget.remove(player);
                    onDeactivate(event);
                }
                else {
                    PlayersUsingGadget.add(player);
                    onActivate(event);
                }
            }

            else {
                if (!PlayersUsingGadget.contains(player)) {
                    RemoveOneItemInHand(player);
                    PlayersUsingGadget.add(player);
                    onActivate(event);

                    new java.util.Timer().schedule(
                            new java.util.TimerTask() {
                                @Override
                                public void run() {
                                    PlayersUsingGadget.remove(player);
                                    onDeactivate(event);
                                }
                            },
                            duration * 1000
                    );
                }
            }
        }
    }

    void RemoveOneItemInHand(Player player) {
        ItemStack item = player.getItemInHand();

        if (item.getAmount() == 1) {
            player.getInventory().setItemInHand(null);
        }
        else {
            player.getInventory().setItemInHand(instance(item.getAmount() - 1));
        }
    }

    public ItemStack instance(int amount) {
        ItemStack stack = this.clone();
        stack.setAmount(amount);
        return stack;
    }

    public ItemStack instance() {
        ItemStack stack = this.clone();
        stack.setAmount(1);
        return stack;
    }
}

File: MinecraftPvpPlugin/src/main/java/me/minecraft/minecraftpvpplugin/Gadget.kt
package me.minecraft.minecraftpvpplugin

import me.minecraft.minecraftpvpplugin.helpers.RunAfter
import me.minecraft.minecraftpvpplugin.refs.Worlds
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.player.PlayerChangedWorldEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.ItemStack

// Duration will only take an effect when SwitchLike = false

// Switch Like:
//     false: player click -> activate() -> delete item -> wait for `duration` s  -> deactivate()
//     true: player click -> activate() -> player click -> deactivate() -> delete item

abstract class Gadget(
    val material: Material,
    private val name: String,
                      private val lore: List<String> = listOf(),
                      private val duration: Double? = null,
                      private val switchLike: Boolean = false,
    ) : ItemStack(), Listener {

        // Can be used to store information about individual players that is activating this gadget
        private var playersActivatingData = mutableMapOf<Player, MutableMap<String, Any>>()

        open fun onActivate(event: PlayerInteractEvent) {}
        open fun onDeactivate(event: PlayerInteractEvent) {}
        open fun onGameEnd(event: PlayerChangedWorldEvent) {}

        init {
            this.type = material
            this.amount = 1

            val meta = this.itemMeta
            meta.displayName = name
            meta.lore = lore
            this.setItemMeta(meta)

            if (!switchLike && duration == null) {
                throw IllegalArgumentException("Duration must be specified if switchLike = false")
            }

            Bukkit.getPluginManager().registerEvents(this, MinecraftPvpPlugin.instance)
        }

        fun addPlayerData(player: Player, key: String, data: Any) {
        playersActivatingData[player]?.set(key, data)
    }

    fun getPlayerData(player: Player, key: String): Any? {
        return playersActivatingData[player]?.get(key)
    }

    fun isActivating(player: Player): Boolean {
        return playersActivatingData.contains(player)
    }

    @EventHandler
    fun handleClick(event: PlayerInteractEvent) {
        val player = event.player
        val item = event.item

        // TODO: Make if nest more readable

        if (Worlds.isInPvp(player) && item?.itemMeta?.displayName == name && event.action != Action.PHYSICAL) {
            if (switchLike) {
                if (isActivating(player)) {
                    removeOneItemInHand(player)
                    onDeactivate(event)
                    playersActivatingData.remove(player)
                }
            } else {
                if (!isActivating(player)) {
                    removeOneItemInHand(player)
                    waitToDeactivate(player, event)
                }
            }

            if (!isActivating(player)) {
                playersActivatingData[player] = mutableMapOf()
                onActivate(event)
            }
        }
    }

    @EventHandler
    fun handleChangeWorld(event: PlayerChangedWorldEvent) {
        if (event.player.world === Bukkit.getWorld("Lobby")) {
            onGameEnd(event)
            playersActivatingData.remove(event.player)
        }
    }

    private fun removeOneItemInHand(player: Player) {
        val item = player.itemInHand

        if (item.amount == 1) {
            player.inventory.itemInHand = null
        } else {
            item.amount -= 1
        }
    }

    private fun waitToDeactivate(player: Player, event: PlayerInteractEvent) {
        RunAfter(duration!!) {
            if (playersActivatingData.contains(player)) {
                onDeactivate(event)
                playersActivatingData.remove(player)
            }
        }
    }
}
